doc/developer/design/20230214_transform_using.md

```diff
@@ -0,0 +1,274 @@
+
+Running dataflows in storage has some unique characteristics which are important to keep in mind
+when evaluating the various options laid out in the rest of the document so I will describe it here
+briefly.
+
+First, some definitions. We will represent the data in the external system that we want to ingest as
+a remote differential collection `RAW`. Storage is tasked with producing a durable storage
+collection `C = PERSIST(TRANSFORM(RAW))` by scheduling executions of ingestion dataflows. Currently
+`TRANSFORM` includes things like format decoding and envelope processing. For each remote collection
+storage maintains a read cursor in it at some frontier `F` that allows it to read the updates to
+`RAW` that happened at times `t: F <= t`. Ingestion dataflows are fault-tolerant and can be
+re-executed when one fails. Executions share no state between them except for any data durably
+stored in `persist` shards. Each execution is tasked with resuming the production of updates to `C`
+by only observing updates to `RAW` that happened beyond `F`.
+
+The unique characteristic is that `RAW` collections do not offer the ability to read a snapshot at
+an arbitrary time `t`, even if that time is beyond our read cursor. `RAW` collections are only able
+to provide the updates that happened at `t` and it is up to the reader to remember past values if
+that is important. The reason for that is that we want to be giving permission to upstream systems
+to *delete* (not just compact) data that we have already seen, if desired.
+
+As a result, the `TRANSFORM` function cannot rely on any in-memory state since its current execution
+can be interrupted at any moment and restarted in a new machine with a blank in-memory state. This
+limits the kinds of dataflow operators that we can use. Stateless operators like `map` and `filter`
+are safe and so are operators that can re-construct their in-memory state by reading a snapshot of
+`C` (not `RAW`). The upsert envelope is an instance of the latter case.
+
+### Adding arbitrary SQL to `TRANSFORM`
+
+Given the constraints of the transformation function that can be ran during ingestion we can explore
+a few different avenues to incorporate an arbitrary query `Q` as a transformation step. Every query
+in materialize, after planning and optimizing, can be decomposed into a `MapFilterProject` (MFP from
+now on) part (`Q_mfp`) and a relation expression part (`Q_expr`). Depending on the query any of
+these can be trivial (i.e the identity).
+
+This decomposition is relevant to what we’re trying to accomplish because `Q_mfp` has the property
+that it’s stateless (state for temporal filters discussed below) and can be evaluated without having
+access to input snapshots whereas `Q_expr` might include arrangements and therefore will need an
+input snapshot before it can resume.
+
+With this decomposition we can instantiate two dataflows:
+
+1. `C_intermediate = PERSIST(Q_mfp(RAW))`
+2. `C = PERSIST(Q_expr(C_intermediate))`
```

Comment:
dunno if anything needs to be done in response to these, but when I saw this, I had two thoughts:

- first, `PERSIST(Q_expr(C_intermediate))` is a definite computation on definite inputs, and so feels very compute-y. possibly odd for this to be living inside the domain of storage. does it have UX consequences for the user (e.g. can they use the same introspection/debugging tools for this dataflow that they'd use if we were doing the same thing but inside a `CREATE MATERIALIZED VIEW`?)
- second, one thing that I've heard as an argument for using `TRANSFORM USING` over something like [pushing down filters into persist reads](https://github.com/MaterializeInc/materialize/issues/12684) is that the former allows a user control over how much data lives in s3 (a cost we pass through to them). this weakens that a bit. I see you and Frank had a discussion above about the benefits to accepting `TRANSFORM USING` queries that can only partially be pushed before persist. given that, how might we give users a mental model for "given X data coming in and a TRANSFORM USING Y, I'll end up paying Z for s3 costs"?

Reply:
I have similar concerns as @danhhz's second point - if the primary product motivation for `TRANSFORM USING` is pushing down temporal filters, the storage cost for a source with temporal filters should be proportional to that. This most probably implies that we need to automatically cleanup `C_intermediate` to keep its size bounded (I suggest to sync with product regarding that, I might be mistaken). Do you have any thoughts how this might happen?

Reply:
These are both good points and it's an omission on my part that I didn't spell out the alternative formulation of these split in half queries. Given my current understanding it seems unlikely that we'll go down the path of supporting them but I comment on what the approach could look like for posterity.

As Dan observed, `Q_expr(C_intermediate)` is a definite computation on definite inputs, and while in this document I always enclosed it in `PERSIST(.)` and gave it a name, it doesn't have to. The alternative path if we ever support split queries would be to compute and store `C_intermediate` as described, but defer the rest of the computation for later. This would mean the source was used in two separate queries `Q1`, `Q2`, we would attach the deferred query fragment before optimizing them and end up with `Q1 = PERSIST(optimize(Q1_expr(Q_expr(C_intermediate))))` and `Q2 = PERSIST(optimize(Q2_expr(Q_expr(C_intermediate))))`. This would have the effect of pushing the maximum MFP down to persist at read time with the trade-off being the re-evaluation of `Q_expr` in every use.

Reply:
Sorry,  I should have put my comment on I think I misplaced my comment on the wrong part of the doc, as we're leaning towards this solution
```text
C_intermediate = PERSIST(Q_expiration(Q_safemfp(RAW)))
C = PERSIST(Q_expr(Q_final(C_intermediate)))
```
In this case, is there a mechanism to keep the size of `C_intermediate`. How are we going to clean up records that have already expired from `C_intermediate`?

Reply:
oh, `Q_expiration` does exactly that! It retracts records at the end of the validity period and because it executes before the first `PERSIST` call the size of `C_intermediate` will be bounded